jmclient/jmclient/wallet.py

```diff
@@ -1013,6 +979,381 @@ def close(self):
+        in PSBT_IN_NONWITNESS_UTXO, PSBT_IN_WITNESS_UTXO and PSBT_IN_REDEEM_SCRIPT
+        can also be provided, one item per input, in the tuple (spent_outs).
+        These objects should be either CMutableTransaction, CTxOut or None,
+        Note that redeem script information cannot be provided for inputs which
+        we don't own.
+        """
+        # TODO: verify tx contains no signatures as a sanity check?
+        new_psbt = btc.PartiallySignedTransaction(unsigned_tx=tx)
+        if spent_outs is None:
+            # user has not provided input script information; psbt
+            # will not yet be usable for signing.
+            return new_psbt
+        for i, txinput in enumerate(new_psbt.inputs):
+            if spent_outs[i] is None:
+                # as above, will not be signable in this case
+                continue
+            if isinstance(spent_outs[i], (btc.CTransaction, btc.CTxOut)):
+                # note that we trust the caller to choose Tx vs TxOut as according
+                # to non-witness/witness:
+                txinput.utxo = spent_outs[i]
+            else:
+                assert False, "invalid spent output type passed into PSBT creator"
+        # we now insert redeemscripts where that is possible and necessary:
+        for i, txinput in enumerate(new_psbt.inputs):
+            if isinstance(txinput.utxo, btc.CMutableTxOut):
+                # witness
+                if txinput.utxo.scriptPubKey.is_witness_scriptpubkey():
+                    # nothing needs inserting; the scriptSig is empty.
+                    continue
+                elif txinput.utxo.scriptPubKey.is_p2sh():
+                    try:
+                        path = self.script_to_path(txinput.utxo.scriptPubKey)
+                    except AssertionError:
+                        # this happens when an input is provided but it's not in
+                        # this wallet; in this case, we cannot set the redeem script.
+                        continue
+                    privkey, _ = self._get_priv_from_path(path)
+                    txinput.redeem_script = btc.pubkey_to_p2wpkh_script(
+                        btc.privkey_to_pubkey(privkey))
+        return new_psbt
+
+    def sign_psbt(self, in_psbt, with_sign_result=False):
+        """ Given a serialized PSBT in raw binary format,
+        iterate over the inputs and sign all that we can sign with this wallet.
+        NB IT IS UP TO CALLERS TO ENSURE THAT THEY ACTUALLY WANT TO SIGN
+        THIS TRANSACTION!
+        The above is important especially in coinjoin scenarios.
+        Return: (psbt, msg)
+        msg: error message or None
+        if not `with_sign_result`:
+        psbt: signed psbt in binary serialization, or None if error.
+        if `with_sign_result` True:
+        psbt: (PSBT_SignResult object, psbt (deserialized) object)
+        """
+        try:
+            new_psbt = btc.PartiallySignedTransaction.from_binary(in_psbt)
+        except:
```

Comment:
Note that you could additionally catch `ValueError` that would give some info on why the operation failed

Reply:
From looking at the `deserialize` method for that class, it seems like there are a few different exception types possible, so I've just caught and printed repr() for general Exception in  https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/9e9ae9e287e2d6e97ea9d53ca628d79fd8e81453 - seems logical?

Reply:
Right, I just meant that `except ValueError:` can be _additional_ case to genral `except:` because `ValueError`s bear some meaningful info about what is wrong with the value (the serialized PSBT)

Reply:
Understood, I just felt I may as well pass through any available error message here, just in case it's useful for debugging (after having read through the many different Serialization/Assertion/Value Errors that can be raised in the library). I guess I would add specific code if there was some action to be performed other than just show the error message.